lsp/src/jsonrpc.ml

```diff
@@ -323,3 +329,88 @@ module Response = struct
 
   let error id error = make ~id ~result:(Error error)
 end
+
+type packet =
+  | Request of Request.t
+  | Response of Response.t
+
+module Session (Chan : sig
+  type t
+
+  val send : t -> packet -> unit Fiber.t
+
+  val recv : t -> packet Fiber.t
+
+  val close : t -> unit Fiber.t
+end) =
+struct
+  type t =
+    { chan : Chan.t
+    ; on_request : Request.t -> Response.t Fiber.t
+    ; on_notification : Request.t -> unit Fiber.t
+    ; pending : (Id.t, Response.t Fiber.Ivar.t) Table.t
+    ; stop_requested : unit Fiber.Ivar.t
+    }
+
+  let on_request_fail (req : Request.t) : Response.t Fiber.t =
+    let id = Option.value_exn req.id in
+    let error =
+      Response.Error.make ~code:InternalError ~message:"not implemented" ()
+    in
+    Fiber.return (Response.error id error)
+
+  let stop t = Fiber.Ivar.fill t.stop_requested ()
+
+  let fork_and_race (type a b) (_ : unit -> a Fiber.t) (_ : unit -> b Fiber.t) :
+      (a, b) Either.t Fiber.t =
+    assert false
+
+  let run t =
+    let rec loop () =
+      let open Fiber.O in
+      let* res =
+        fork_and_race
+          (fun () -> Chan.recv t.chan)
+          (fun () -> Fiber.Ivar.read t.stop_requested)
+      in
+      match res with
+      | Either.Right () -> Chan.close t.chan
+      | Left (Request r) ->
+        let* resp = t.on_request r in
+        let* () = Chan.send t.chan (Response resp) in
+        loop ()
+      | Left (Response r) -> (
+        match Table.find t.pending r.id with
+        | None -> Code_error.raise "no pending request" []
```

Comment:
I wouldn't trust a server/client to always be correct. Maybe it should just log and continue if it receives a response that does not correspond to a request ?

Reply:
Okay, I ignored it. It doesn't mater so much for now since we'll be dealing with well behaved clients/servers, so let's just do what's easier.